### PR TITLE
use pinned deps to specify sysimage contents

### PR DESCRIPTION
**Pull request by kolia**
see #4

This was easier than anticipated; I just put the extraction of pinned deps in yet another docker stage.

README is updated to reflect new convention.

### CODE REVIEW

File: add_me_to_your_PATH/build_image [hunk 1]
```diff
@@ -78,12 +76,11 @@ bash $SCRIPT_DIR/tight_deps.sh
 mkdir -p dev
 
 VERSIONS_TAG="cuda-${CUDA_VERSION}_julia-${JULIA_VERSION}"
-SYSIMAGE_DEPS_SHA="sysimage-$(sha1sum TightProject.toml | head -c7)"
 DEPS_SHA="precompiled-$(sha1sum Project.toml Manifest.toml | sha1sum | head -c7)"
 
 # CACHE TAGs, designed to only change if cache images change (can we get a hash from the image itself instead?)
 BASE_IMAGE_CACHE="${IMAGE_REPO}:baseimage_${VERSIONS_TAG}"
```
Comment (omus) [inline add_me_to_your_PATH/build_image:82]: May as well keep the naming convention similar
```suggestion
BASE_IMAGE_CACHE="${IMAGE_REPO}:${VERSIONS_TAG}_base"
```

File: add_me_to_your_PATH/sysimage.jl [hunk 2]
```diff
@@ -18,11 +18,9 @@ Pkg.add("PackageCompiler")
 using PackageCompiler, UUIDs
 
 function main()
-    packages = [Symbol(k) for k in keys(project.dependencies)]
-    # exclude dependencies that are being developed
-    exclude = [Symbol(pi.name) for pi in values(Pkg.dependencies()) if !pi.is_tracking_registry]
+    packages = [Symbol(k) for k in keys(Pkg.project().dependencies)]
```
Comment (omus) [inline add_me_to_your_PATH/sysimage.jl:21]: Why was this line changed? Should the portion code before this be placed in this `main` function?

File: add_me_to_your_PATH/sysimage.jl [hunk 3]
```diff
@@ -18,11 +18,9 @@ Pkg.add("PackageCompiler")
...
     # exclude some packages that can make sysimage creation or usage fail.
-    foreach(d -> push!(exclude, d), [:Revise, :PackageCompiler, :K8sClusterManagers])
+    exclude = [:Revise, :PackageCompiler, :K8sClusterManagers]
```
Comment (omus) [inline add_me_to_your_PATH/sysimage.jl:23]: Do we still need to exclude K8sClusterManagers?

File: README.md [hunk 4]
```diff
@@ -135,21 +135,24 @@ The default `Dockerfile` used by `julia_pod` is set up to
...
 
 The convention for inclusion into the sysimage is that any
-project dependency with a **tight** compat bound is included,
+project dependency that is **pinned** included,
 unless it is blacklisted as a package known not to work with
-`PackageCompiler.jl`.
+`PackageCompiler.jl`. To add dependency `TheDependency` to the
+sysimage, in a julia REPL `Pkg` mode, do `]pin TheDependency`.
+Pinned packages show with a little ⚲ next to them when listed
+with `]status`.
```
Comment (omus) [inline README.md:143]:
```suggestion
`PackageCompiler.jl`. To add the dependency `Example` to the
sysimage, in a julia REPL `Pkg` mode, do `]pin Example`.
Pinned packages show with a little ⚲ next to them when listed
with `]status`.
```

File: README.md [hunk 5]
```diff
@@ -135,21 +135,24 @@ The default `Dockerfile` used by `julia_pod` is set up to
...
 
 This convention came about because the package manager `Pkg.jl`
 is not aware of what dependencies are included in the sysimage
 (they are treated as un-versioned dependencies as though they
 were part of the julia std lib, and always take precedence over
```
Comment (omus) [inline README.md:148]:
```suggestion
were part of the julia stdlib, and always take precedence over
```

Comment (omus) [reply README.md:148]: FYI: In the official documentation Julia refers to the language while `julia` refers to the executable. I typically use "Julia" in documentation

File: README.md [hunk 6]
```diff
@@ -135,21 +135,24 @@ The default `Dockerfile` used by `julia_pod` is set up to
...
 whatever dependencies `Pkg.jl` thinks it has added to the
-project): it is good practice to place tight bounds on any
+project): it is good practice to pin the versions of
 dependencies that are in a sysimage, so that the versions that
 are actually being used match the versions that `Pkg.jl` thinks
 it is working with. This is why we decided to just make the set
```
Comment (omus) [inline README.md:153]:
```suggestion
project). Pinning the packages that are included in the sysimage 
ensures that the package versions used by `Pkg.jl` are fixed 
and cannot be accidentally changed. Without pinning a user
could change the version of a package via Pkg but find that they
are still only using the version locked into the sysimage.
This is why it was decided to use pinned packages to control
what is built into the sysimage.
```

File: README.md [hunk 7]
```diff
@@ -135,21 +135,24 @@ The default `Dockerfile` used by `julia_pod` is set up to
...
-of dependencies that have tight bounds be what gets included
-in sysimages by default.
+of dependencies that have pinned versions be those that get
+included in the built sysimage.
```
Comment (omus) [inline README.md:155]:
```suggestion
```

**Comment by omus** [comment]
@kolia great use of a multi-stage build for this :italian-chef-kiss:

File: add_me_to_your_PATH/sysimage.jl (hunk 3 above)
Comment (kolia) [reply add_me_to_your_PATH/sysimage.jl:23]: Not sure, haven't tried in a while, the new version might not have issues. Lemme give it a shot.

File: add_me_to_your_PATH/sysimage.jl (hunk 2 above)
Comment (kolia) [reply add_me_to_your_PATH/sysimage.jl:21]: Not sure why but if I put the `const project = ...` and following blocks in `main`, I get a

```
LoadError: expected the file `src/Example.jl` to exist for package `Example` at `/JuliaProject`
```.

I'm just reverting to using the `const project` in `main` as it was before.

File: add_me_to_your_PATH/sysimage.jl (hunk 3 above)
Comment (kolia) [reply add_me_to_your_PATH/sysimage.jl:23]: I just tried with Revise and K8sClusterManagers both pinned / in the sysimage and it's working. Removing both from excludes.

File: add_me_to_your_PATH/sysimage.jl (hunk 2 above)
Comment (omus) [reply add_me_to_your_PATH/sysimage.jl]: You cant define `const`ants in functions

Comment (kolia) [reply add_me_to_your_PATH/sysimage.jl]: Oh I had removed the `const` keywords when I tried putting them in `main`.

Not sure what is causing the LoadError when I try to do in in `main`, but writing an empty `src/Example.jl` outside of `main` seems to do the trick.

**Review by omus — APPROVED** [review]
